extensions/ipynb/src/notebookImagePaste.ts

```diff
@@ -0,0 +1,145 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as vscode from 'vscode';
+
+class CopyPasteEditProvider implements vscode.DocumentPasteEditProvider {
+
+	async provideDocumentPasteEdits(
+		_document: vscode.TextDocument,
+		_ranges: readonly vscode.Range[],
+		dataTransfer: vscode.DataTransfer,
+		_token: vscode.CancellationToken
+	): Promise<vscode.DocumentPasteEdit | undefined> {
+
+		// get foundational notebook data
+		const activeNotebook = vscode.window.activeNotebookEditor?.notebook;
+		const activeSelection = vscode.window.activeNotebookEditor?.selection;
+		const currentCell = activeNotebook?.getCells(activeSelection);
+		if (!currentCell || currentCell?.length !== 1) {
+			return undefined;
+		}
+		const initialAttachments = currentCell[0].metadata?.custom?.attachments;
+		const initialize = !initialAttachments ? true : false;
+
+		// get b64 data from paste
+		const dataItem = dataTransfer.get('image/png');
+		if (!dataItem) {
+			return undefined;
+		}
+		const fileDataAsUint8 = await dataItem.asFile()?.data();
+		if (!fileDataAsUint8) {
+			return undefined;
+		}
+		const b64string = encodeBase64(fileDataAsUint8);
+
+		// get filename data from paste
+		let pasteFilename = dataItem.asFile()?.name;
+		const filePrefix = pasteFilename?.split('.')[0];
+		const fileSuffix = pasteFilename?.split('.')[1];
+		if (!filePrefix || !fileSuffix) {
+			return undefined;
+		}
+
+		// create updated metadata for cell (prep for WorkspaceEdit)
+		if (!initialize) {
```

Comment:
Do you need to iterate though all attachments or can you do something like:

- Is `my-image` an open name?
- If not, is `my-image-2` an open name?
- If not, is `my-image-3` an open name? 
-  ...

Reply:
will be fixed in next commit. changed name validity checking to a while loop + `filename in initialAttachments` check